### DISCUSSION TITLE
Access / add attributes in a filter

### DISCUSSION THREAD

**Discussion by pagiraud** [jgm/pandoc#10177] [Q&A] [answered]
Hi,

I’m making a filter for revealJS presentations and I’m struggling with modifying attributes of block elements. Actually (this is just a portion of my code that I think is sufficient to understand my question, but if needed I can post the full code or even the AST), I don’t understand why this works (ie the background becomes blue):

``` lua
local function bblocks (blocks)
  local figuretype
  -- Go from end to start to avoid problems with shifting indices.
  for i = #blocks-1, 1, -1 do
    figuretype = is_figure_in_fullframegraphic_frame(blocks[i], blocks[i+1])
    if figuretype == "Figure" then
      blocks[i].attributes.style = "background-color: blue"
    elseif figuretype == "Para" then
      print("Para")
    end
  end
  return blocks
end
```
But that doesn’t (note the line with background-image, error below):

```lua
local function bblocks (blocks)
  local figuretype
  -- Go from end to start to avoid problems with shifting indices.
  for i = #blocks-1, 1, -1 do
    figuretype = is_figure_in_fullframegraphic_frame(blocks[i], blocks[i+1])
    if figuretype == "Figure" then
      blocks[i].attributes.background-image = "img.jpg"
    elseif figuretype == "Para" then
      print("Para")
    end
  end
  return blocks
end
```

Error:

```
ERROR (/opt/quarto/share/filters/main.lua:3868) ...path/to/ffgreveal.lua:35: syntax error near '-'
```

Thanks for the help!

**Comment by tarleb** [comment, marked as answer]
Lua allows to use `my_table.key` as an alias for `my_table['key']`, i.e., it looks `key` up in table `my_table`. However, this works only if the key is a *Name*, which the Lua manual defines like this:

> *Names* (also called *identifiers*) in Lua can be any string of Latin letters, Arabic-Indic digits, and underscores, not beginning with a digit and not being a reserved word. Identifiers are used to name variables, table fields, and labels.

Thus, `background-image` is not a name, which means that square brackets have to be used:

``` lua
blocks[i].attributes['background-image'] = "img.jpg"
```

**Reply by pagiraud** [reply]
Thanks, that makes sense and that works! 

Edit: Maybe, in the documentation about LUA filters, a good idea would be to include such an example, because the only example available uses `attributes.style`, and that’s why I tried this way.
